modules/gateway/peers.go

```diff
@@ -326,6 +351,36 @@ func connectVersionHandshake(conn net.Conn, version string) (remoteVersion strin
 	return remoteVersion, nil
 }
 
+// connectSessionHandshake performs the sessionHeader handshake and should be called
+// on the side making the connection request, whether that's for pure pinging reasons or not.
+// a connection is only possible if err == nil
+func connectSessionHandshake(conn net.Conn, header sessionHeader) (wantConn bool, err error) {
+	// Send our header.
+	if err = encoding.WriteObject(conn, header); err != nil {
+		err = fmt.Errorf("failed to write header: %v", err)
+		return
+	}
+	// Read remote header.
+	var remoteHeader sessionHeader
+	if err = encoding.ReadObject(conn, &remoteHeader, build.MaxEncodedVersionLength); err != nil {
+		err = fmt.Errorf("failed to read remote header: %v", err)
+		return
+	}
+	// ensure that their genesisID is the same as ours
+	if bytes.Compare(header.GenesisID[:], remoteHeader.GenesisID[:]) != 0 {
+		err = errPeerRejectedConn
+		return
+	}
+	// ensure that we're not connecting to ourselves
+	if bytes.Compare(header.UniqueID[:], remoteHeader.UniqueID[:]) == 0 {
+		err = errPeerExists
+		return
+	}
+	// return if both sesions want a connection
+	wantConn = header.WantConn && remoteHeader.WantConn
```

Comment:
I changed this behavior back and forward quite a few times during development, sometimes returning exactly what you're now asking. Eventually I settled on returning the header. But I think you're right, there is never really any need to know the other information after we've settled the information is matching.